### ISSUE TITLE
Prevent inserting <object> into <head>

### ISSUE THREAD

**Issue by emerick** [htacg/tidy-html5#232]
Given the following piece of HTML:

```
<object class="todo"></object>
```

HtmlTidy will insert the ```<object>``` into ```<head>```, like this:

```
<!DOCTYPE html>
<html>
<head>
<meta name="generator" content=
"HTML Tidy for HTML5 for Windows version 5.0.0.RC1">
<object class="todo"></object>
<title></title>
</head>
<body>
</body>
</html>
```

Given the limited context of the input, inserting ```<object>``` into ```<body>``` seems like a smarter default. Or perhaps an option can be added to HtmlTidy to decide how to handle this case.

**Comment by geoffmcl** [comment]
@emerick, hmmm, an interesting case... thanks for reporting...

Yes, internally Tidy! marks the `<object>` element as being allowed in `<head>`, and on being given no other clues as to what the user wants, chooses to put it in an `implied` head...

Of course, normally an `<object>` element would be part of another block, like -

```
<p>See my video
<object type="application/x-shockwave-flash">
<param name=movie value="http://example.com/watch.swf">
</object>
</p>
```

in which case Tidy! would put it in the `<body`, since `<p>` is not allowed in the head... and `<object>` is a child of `<p>` so goes with it...

So, I do not feel this is a **bug** ;=))

Now, if important enough, then yes we could added some additional option to tell Tidy! to use `<body>` in this `<object>` case...

And a decision made whether this **new** option, when implemented, would effect other specific elements, like say -

```
<script>
var val = 0;
</script>
```

which again, without other user clues, will be placed in the `<head>`...

Personally, I am not sure such a **new** option is warranted, given the **very** limited case, but will leave this open for a while for further comments... and marking this as a `feature` with an indefinite milestone...

**Comment by emerick** [comment]
@geoffmcl Thanks very much for the feedback. The main reason I think that this could be a compelling addition is that HTML5 no longer accepts ```<object>``` tags inside ```<head>```. This means that going forward, placing ```<object>``` in ```<body>``` is a much more sensible default, but there will still be a need for flexibility for older documents.

I do agree that it's a niche case, I'm just not convinced that ```<head>``` is the best possible guess with such limited context.

**Comment by geoffmcl** [comment]
@emerick, oops, sorry, the one W3C ref I consulted did not seem to mention not in head in html5... maybe you can give me some urls, refs? Just to confirm this...

If this **is** the case then the fix looks trivial ;=))

We already have several cases where Tidy! is internally setup to process html5++. But if it detects a previous doctype, will change some structures, values, flags to process the document as html4--.

In `void TY_(AdjustTags)( TidyDocImpl *doc )` we already adjust TidyTag_A - Issue #167 & #169, and TidyTag_CAPTION - Issue #196... so this would just be an addition...

The `TidyTag_OBJECT` line in `tag_defs` would have the `CM_HEAD` bit removed for the default html5++ mode, and then your html snippet should be placed in the implied body... 

But if html4-- detected, then `or` this bit back into the table...

Yet to get around to testing this... maybe you will beat me to it ;=))

**Comment by geoffmcl** [comment]
@emerick, This patch would allow you to test...

```
diff --git a/src/tags.c b/src/tags.c
index 5470831..5e34ead 100644
--- a/src/tags.c
+++ b/src/tags.c
@@ -160,8 +160,8 @@ static CheckAttribs CheckHTML;
 #define VERS_ELEM_WBR        (xxxx|xxxx|xxxx|xxxx|xxxx|xxxx|xxxx|xxxx|xxxx|xxxx|xxxx|xxxx|xxxx|HT50|XH50)
 
 /*\ 
- * Issue #167 & #169
- * Tody defaults to HTML5 mode
+ * Issue #167 & #169 & #232
+ * Tidy defaults to HTML5 mode
  * but allow this table to be ADJUSTED if NOT HTML5
  * was static const Dict tag_defs[] = 
 \*/
@@ -232,7 +232,7 @@ static Dict tag_defs[] =
   { TidyTag_META,       "meta",       VERS_ELEM_META,       &TY_(W3CAttrsFor_META)[0],       (CM_HEAD|CM_BLOCK|CM_EMPTY),                   TY_(ParseEmpty),    NULL           },
   { TidyTag_NOFRAMES,   "noframes",   VERS_ELEM_NOFRAMES,   &TY_(W3CAttrsFor_NOFRAMES)[0],   (CM_BLOCK|CM_FRAMES),                          TY_(ParseNoFrames), NULL           },
   { TidyTag_NOSCRIPT,   "noscript",   VERS_ELEM_NOSCRIPT,   &TY_(W3CAttrsFor_NOSCRIPT)[0],   (CM_HEAD|CM_BLOCK|CM_INLINE|CM_MIXED),         TY_(ParseBlock),    NULL           },
-  { TidyTag_OBJECT,     "object",     VERS_ELEM_OBJECT,     &TY_(W3CAttrsFor_OBJECT)[0],     (CM_OBJECT|CM_HEAD|CM_IMG|CM_INLINE|CM_PARAM), TY_(ParseBlock),    NULL           },
+  { TidyTag_OBJECT,     "object",     VERS_ELEM_OBJECT,     &TY_(W3CAttrsFor_OBJECT)[0],     (CM_OBJECT|CM_IMG|CM_INLINE|CM_PARAM),         TY_(ParseBlock),    NULL           },
   { TidyTag_OL,         "ol",         VERS_ELEM_OL,         &TY_(W3CAttrsFor_OL)[0],         (CM_BLOCK),                                    TY_(ParseList),     NULL           },
   { TidyTag_OPTGROUP,   "optgroup",   VERS_ELEM_OPTGROUP,   &TY_(W3CAttrsFor_OPTGROUP)[0],   (CM_FIELD|CM_OPT),                             TY_(ParseOptGroup), NULL           },
   { TidyTag_OPTION,     "option",     VERS_ELEM_OPTION,     &TY_(W3CAttrsFor_OPTION)[0],     (CM_FIELD|CM_OPT),                             TY_(ParseText),     NULL           },
@@ -762,6 +762,20 @@ void TY_(AdjustTags)( TidyDocImpl *doc )
         tagsEmptyHash( doc, tags );
 #endif
     }
+
+/*\
+ * Issue #232
+ * TidyTag_OBJECT not in head in HTML5,
+ * but still allowed in HTML4
+\*/
+    np = (Dict *)TY_(LookupTagDef)( TidyTag_OBJECT );
+    if (np)
+    {
+        np->model |= CM_HEAD; /* add back allowed in head */
+#if ELEMENT_HASH_LOOKUP
+        tagsEmptyHash( doc, tags );
+#endif
+    }
 }
 
 void TY_(FreeTags)( TidyDocImpl* doc )
```

Looks fine in initial tests...

**Comment by emerick** [comment]
@geoffmcl Thanks for the fast turnaround on this patch, I really appreciate it! I tested it this morning and it works great for my use case and the changes make sense to me.

Regarding your question about where this is called out in the spec, this section of the spec shows what links qualify as "metadata content" (which is all that's allowed in head, as of HTML5):

http://www.w3.org/TR/html5/dom.html#metadata-content-0

You'll see that object is not listed under "3.2.4.1.1 METADATA CONTENT", but is listed under "3.2.4.1.2 FLOW CONTENT". In addition, a document with an object in head doesn't validate via the W3C validator when "HTML5 Experimental" is selected.

**Comment by geoffmcl** [comment]
@emerick, I am presently working in the issue-228 branch, which includes a new cool feature, see #231 and #228, so have added this 'fix' there. You can test it with -

$ git checkout issue-228
$ git pull  # if above done previously
$ cd build/cmake
$ cmake-clean # at least delete CMakeCache.txt
$ cmake ../.. [options]
$ make

Checking the reference you gave, and using the 6 MetaData content give of `base link meta noscript script style template title` **Tidy!** does not yet include `template`!

It means adjusting the following template table line to include the CM_HEAD bit...

```
  { TidyTag_TEMPLATE,    "template",     VERS_ELEM_TEMPLATE,    &TY_(W3CAttrsFor_TEMPLATE)[0],    (CM_BLOCK),                    TY_(ParseBlock),     NULL           },
```

Perhaps this should be part of this fix also? Or perhaps **NOT**? Undecided on this, and comments welcome.

And just out of interest found the CM_HEAD bit is in the following tags `"nextid", "bgsound", "server", "command"` in **Tidy!**. Have not yet figured out all the ramifications of this, and whether anything else should be changed. Again comments welcome.

There have been a few weeks of testing of this `issue-228` branch, and will shortly, maybe another week or so, be merging it all back into the development branch, `master`...

Meantime appreciate any testing and feedback...

**Comment by emerick** [comment]
@geoffmcl Your patch works great on the issue-228 branch, thanks for getting it in there.

Regarding adding support for ```template```: I can't say that I'm too familiar with it myself, but anything we can do to increase compliance seems good (although you may want to err on the side of caution until someone specifically calls it out as an issue).

**Comment by geoffmcl** [comment]
@emerick, just a small note to advise the `issue-228` branch has now been fully merged into the `master` branch, from version 5.1.8 (2015.09.04) onwards... and will at some stage delete the `issue-228` development branch...

It would be great if you get a chance to update to `master` and rebuild -

```
$ cd tidy-html5
$ git checkout master
$ git pull
```

I have re-tested this bug with 5.1.8 and it seems/remains fixed, so am closing this. Please feel free to re-open, or open another issue...

I agree with the caution on the `template` tag, until someone specifically has a case and opens an issue, adding sample html, and adds W3C HTML5 support references...

Thanks for this little `tweak` ;=))

**Comment by emerick** [comment]
Thanks @geoffmcl!
